sys: timeutil: add utility functions for struct timespec

cfriedt opened the pull request:
Add a number of utility functions for manipulating struct timespec.

* timespec_add()
* timespec_compare()
* timespec_equal()
* timespec_is_valid()
* timespec_negate()
* timespec_normalize()
* timespec_sub()
* timespec_from_timeout()
* timespec_to_timeout()

If the `__builtin_add_overflow()` function is available, then these functions use are mostly branchless, which should provide decent performance on systems with caches and branch prediction (gated by `CONFIG_SPEED_OPTIMIZATIONS`). Otherwise, manually written alternatives exist that are also perhaps more readable.

The two functions at the end convert time durations between representation as `struct timespec` and `k_timeout_t`.

Fixes #88115

[Doc Preview](https://builds.zephyrproject.io/zephyr/pr/90060/docs/kernel/timeutil.html)

Testing Done:
```shell
twister -i -T tests/lib/timespec_util
INFO    - Total complete:   45/  45  100%  built (not run):    0, filtered:  280, failed:    0, error:    0
INFO    - 7 test scenarios (322 configurations) selected, 280 configurations filtered (277 by static filter, 3 at runtime).
INFO    - 42 of 42 executed test configurations passed (100.00%), 0 built (not run), 0 failed, 0 errored, with no warnings in 49.20 seconds.
INFO    - 378 of 378 executed test cases passed (100.00%) on 9 out of total 1058 platforms (0.85%).
INFO    - 42 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

Commit:
tests: lib: timeutil: add timespec util testsuite

Add a timespec util testsuite. This should have reasonably high enough
coverage to be useful.

I would have preferred to add this as an architecture-independent
unit test (for the unit_testing platform) under tests/unit/timeutil but
there is an inconsistency about the size of time_t on the unit_testing
and native_sim/native platforms. On every other platform supported by
Zephyr, time_t is 64-bits. However, on those platforms, time_t is only
32-bits.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: timers: use newly added timespec util functions

Use the newly added timespec util functions to manipulate and
compare timespec structures with overflow detection.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented on tests/lib/timespec_utils/testcase.yaml:
```diff
@@ -0,0 +1,28 @@
+    - timeutils
+  integration_platforms:
+    - native_sim/64
+tests:
```
It would also be good to add a testcase to verify the non-builtin code path.

Keeping this in draft until I've done that.

cfriedt commented:
cc @trembel

bjarki-andreasen commented on include/zephyr/sys/timespec_util.h:
```diff
@@ -0,0 +1,314 @@
+ *
+ * @return `true` if the operation would result in integer overflow, otherwise `false`.
+ */
+static inline bool timespec_add(struct timespec *a, const struct timespec *b)
```
if `timespec` itself is not nameclashing, the header should probably just be named `timespec.h` given the namespace `timespec_util` is not used for any function.

cfriedt replied:
I just merged it directly into `timespec.h`

https://github.com/zephyrproject-rtos/zephyr/pull/90060#discussion_r2093165672

bjarki-andreasen reviewed (commented):
Since timespec is specific to posix, should it be moved to `include/zephyr/posix/` where
```
struct` timespec {
	time_t tv_sec;
	long tv_nsec;
};
```
is also defined? `include/zephyr/sys` I believe is primarily for zephyr native libraries, though there obviously are a few overlaps there like fdtable. Not sure how or where we draw such lines :)

cfriedt commented on include/zephyr/sys/timespec_util.h:
Based on the structure of [the documentation](https://docs.zephyrproject.org/latest/kernel/timeutil.html), it might be better just to add the contents of this file directly to `include/zephyr/sys/timeutil.h`

Commit:
doc: kernel: timeutil: add documentation for timespec apis

Add documentation in the timeutil group for recently added functions
for validating, comparing, and manipulating `struct timespec`
objects as well as functions for converting between representation
of time durations as either `struct timespec` or `k_timeout_t`.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
timeutil: correct position of doxygen end-group comment

swap the relative position of the __cplusplus closing
bracket and the doxygen end-group comment.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
sys: timeutil: add utility functions for struct timespec

Add a number of utility functions for manipulating struct timespec.

* timespec_add()
* timespec_compare()
* timespec_equal()
* timespec_is_valid()
* timespec_negate()
* timespec_normalize()
* timespec_sub()
* timespec_from_timeout()
* timespec_to_timeout()

If the `__builtin_add_overflow()` function is available, then the
API is mostly branchless, which should provide decent performance on
systems with an instruction cache and branch prediction. Otherwise,
manually written alternatives exist that are also perhaps more
readable.

The two functions at the end convert time durations between
representation as `struct timespec` and `k_timeout_t`.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

npitre reviewed (commented):
A few comments:

First, the bool return value. It feels to me like it should be the other
way around i.e. true=OK and false=problem. Currently you have things like:

```
    return timespec_negate(&neg) || timespec_add(a, &neg);
```

This just looks semantically wrong. It should be like: if we can negate
_and_ add then return OK e.g.:

```
    return timespec_negate(&neg) && timespec_add(a, &neg);
```

Another example:

```
    return __builtin_add_overflow(a->tv_sec, b->tv_sec, &a->tv_sec) ||
           __builtin_add_overflow(a->tv_nsec, b->tv_nsec, &a->tv_nsec) ||
           timespec_normalize(a);
```

IMHO this would be more logically readable as:

```
    return !__builtin_add_overflow(a->tv_sec, b->tv_sec, &a->tv_sec) &&
           !__builtin_add_overflow(a->tv_nsec, b->tv_nsec, &a->tv_nsec) &&
           timespec_normalize(a);
```

Translation: if no overflow on the first add _and_ no overflow on the second
add _and_ normalization worked then return OK (true). Yes,
`__builtin_add_overflow()` returns true on error but that's in the name.
Here we have `timespec_add()` and the name doesn't imply an error.
Users would have more readable code if:

```
    if (timespec_add(a, b) {
        /* true: the addition was possible */
    } else {
        /* false: addition was _not_ possible */
    }
```

Next, I'm a little concerned by the optimization. This:

```
    int64_t sec = (ts->tv_nsec >= NSEC_PER_SEC) * (ts->tv_nsec / NSEC_PER_SEC) +
+                 (ts->tv_nsec < 0) * DIV_ROUND_UP(-ts->tv_nsec, NSEC_PER_SEC);
```

That's potentially two divisions just for the purpose of avoiding a branch.
On many architectures (if not most) a branch is still cheaper than a
division, even more so if the division ends up in a library call given
a 64-bit dividend.

And finally, you should add more values to your test case. In particular,
tests with tv_nsec being `-NSEC_PER_SEC - 1`, `-2*NSEC_PER_SEC + 1`,
`-2*NSEC_PER_SEC - 1`, etc. would provide more coverage.

cfriedt commented:
> Since timespec is specific to posix, should it be moved to `include/zephyr/posix/`

@bjarki-andreasen - mentioned in these docs as well as the C standard, while `struct timespec` originated with POSIX, it has been a part of ISO C since C11.

bjarki-andreasen commented:
> > Since timespec is specific to posix, should it be moved to `include/zephyr/posix/`
> 
> @bjarki-andreasen - mentioned in these docs as well as the C standard, while `struct timespec` originated with POSIX, it has been a part of ISO C since C11.
> 
> See also
> https://en.cppreference.com/w/c/chrono/timespec
> 

Fair enough :) I have no strong opinion, just noticed that we do define the timespec struct in the posix folder. Anyway PR looks good, definate improvement!

cfriedt commented:
> IMHO this would be more logically readable as:
> 
> ```
>     return !__builtin_add_overflow(a->tv_sec, b->tv_sec, &a->tv_sec) &&
>            !__builtin_add_overflow(a->tv_nsec, b->tv_nsec, &a->tv_nsec) &&
>            timespec_normalize(a);
> ```

From my perspective, that is worse - there are 5 logical operations instead of 3.

> Yes, `__builtin_add_overflow()` returns true on error but that's in the name. 

I'm sort of thinking I would prefer to just append `_overflow`. To me, it's less useful to add multiple extra logical operations.

> Here we have `timespec_add()`.

Yeah, I could definitely rename that to `timespec_add_overflow()`. That would be in-line with all of the `math-extras.h` functions.

https://github.com/zephyrproject-rtos/zephyr/tree/main/include/zephyr/sys/math_extras.h

> Next, I'm a little concerned by the optimization. This:
> 
> ```
>     int64_t sec = (ts->tv_nsec >= NSEC_PER_SEC) * (ts->tv_nsec / NSEC_PER_SEC) +
> +                 (ts->tv_nsec < 0) * DIV_ROUND_UP(-ts->tv_nsec, NSEC_PER_SEC);
> ```
> 
> That's potentially two divisions just for the purpose of avoiding a branch. On many architectures (if not most) a branch is still cheaper than a division, even more so if the division ends up in a library call given a 64-bit dividend.

Agreed.

What I initially had is only cheap on processors that have e.g. icache, dcache, HW divide, and only cheap compared to potential cache misses.

I would be happy to use `#ifdef CONFIG_SPEED_OPTIMIZATIONS` here, because you are right in pointing out that the usual optimization for Zephyr should be for size.

> And finally, you should add more values to your test case. In particular, tests with tv_nsec being `-NSEC_PER_SEC - 1`, `-2*NSEC_PER_SEC + 1`, `-2*NSEC_PER_SEC - 1`, etc. would provide more coverage.

Absolutely - thanks for the feedback @npitre.

cfriedt commented:
> Fair enough :) I have no strong opinion, just noticed that we do define the timespec struct in the posix folder. Anyway PR looks good, definate improvement!

Yeah, that's true - we should probably eliminate duplicate definitions especially with #30105

npitre reviewed (commented):
> > IMHO this would be more logically readable as:
> >
> > ```
> >     return !__builtin_add_overflow(a->tv_sec, b->tv_sec, &a->tv_sec) &&
> >            !__builtin_add_overflow(a->tv_nsec, b->tv_nsec, &a->tv_nsec) &&
> >            timespec_normalize(a);
> > ```
> 
> From my perspective, that is worse - there are 5 logical operations
> instead of 3.

Absolutely not. There are no more _operations_. The compiler is happy
to flip the logic around as it see fit and it won't care one way or the
other.. OTOH this is better semantics for humans reading the code.

> I'm sort of thinking I would prefer to just append `_overflow`. To me,
> it's less useful to add multiple extra logical operations.

Please don't! This is an abomination!

Adding `_overflow` everywhere instead of `!` in a few places is a really
a bad tradeoff. Again, the compiler doesn't care either ways (the code won't
be any bigger or any less efficient).

cfriedt commented:
> Please don't! This is an abomination!

Hehe.. sure, if you insist. I feel less strongly

ycsin reviewed (commented):
LGTM

cfriedt commented:
* removed `_overflow()` from fn signatures, invert return values, adjust docs
* cast `NSEC_PER_SEC` to `long` to ensure correct behaviour for 32-bit architectures

cfriedt commented:
I might take a crack at fixing #90029 here as well. It would be nice to be able to put the tests in the `tests/unit/timeutil` testsuite when they are initially committed instead of moving them later.

cfriedt commented:
* fixed some additional `(long)` / `(unsigned long)` corner cases for 32-bit `tv_nsec`
* mitigated 2 SonarQube warnings related to MISRA

cfriedt commented:
* decided against trying to use 64-bit `time_t` on `native_sim/native` and `unit_testing` platforms because it is still generally a [known](https://blogs.gentoo.org/mgorny/2024/09/28/the-perils-of-transition-to-64-bit-time_t/) and [not completely solved issue in Linux distributions](https://wiki.debian.org/ReleaseGoals/64bit-time)

cfriedt commented:
* remaining SonarQube issues are false positives (static const array in tests without dimensions)
* removed `__builtin_sub_overflow()` since making a value less negative than `LONG_MIN` or less positive than `LONG_MAX` is guaranteed to never overflow.

cfriedt commented:
* removed unneeded check `HAS_BUILTIN(__builtin_sub_overflow)`
* reworded commit message since `__builtin_sub_overflow()` is no longer needed

cfriedt commented:
* inverted usage of `timespec_normalize()` in `z_clock_gettime()` following changes to `timespec_normalize()`

cfriedt commented:
* removed `BUILD_ASSERT()`s from `timeutil.h`

npitre reviewed (dismissed):
😊

peter-mitsis reviewed (dismissed):
Looks good to me.

All the division in timespec_normalize() did have me concerned for a bit from a performance perspective, but then I realized that we are dividing by a well known constant which most (if not all) compilers should be able to convert into a short set of mutliply and shift operations.

cfriedt commented:
@kartben - in terms of doxygen and I guess code cleanliness, would it be better to use 

```cpp
#if defined(CONFIG_SPEED_OPTIMIZATIONS) || defined(__DOXYGEN__)
/**
 * doxygen comments
 */
static inline void foo(void)
{
  /* bigger, faster */
}
#else
static inline void foo(void)
{
  /* smaller, maybe slower */
}
#endif
```
or something like this?
```cpp
/**
 * doxygen comments
 */
static inline void foo(void)
{
#if defined(CONFIG_SPEED_OPTIMIZATIONS)
  /* bigger, faster */
#else
  /* smaller, maybe slower */
#endif
}
```

kartben commented:
> @kartben - in terms of doxygen and I guess code cleanliness, would it be better to use
> 
> ```c++
> #if defined(CONFIG_SPEED_OPTIMIZATIONS) || defined(__DOXYGEN__)
> /**
>  * doxygen comments
>  */
> static inline void foo(void)
> {
>   /* bigger, faster */
> }
> #else
> static inline void foo(void)
> {
>   /* smaller, maybe slower */
> }
> #endif
> ```
> 
> or something like this?
> 
> ```c++
> /**
>  * doxygen comments
>  */
> static inline void foo(void)
> {
> #if defined(CONFIG_SPEED_OPTIMIZATIONS)
>   /* bigger, faster */
> #else
>   /* smaller, maybe slower */
> #endif
> }
> ```
> 
> Probably good to catch this now, since it gives me an opportunity to also add the following as well as tests.
> 
> ```c++
> static inline k_timeout_t timespec_to_timeout(const struct timespec *ts);
> static inline void timespec_from_timeout(k_timeout_t timeout, struct timespec *ts);
> ```

a preference for the second option :)

cfriedt commented:
* reworked `CONFIG_SPEED_OPTIMIZATIONS` switch so that doxygen is always visible / switch happens inside of function
* added `timespec_from_timeout()` and `timespec_to_timeout()` and tests

Please have another look @npitre @peter-mitsis

npitre reviewed (dismissed):
You are doing:

```
     .tv_sec = (timeout.ticks == K_TICKS_FOREVER) * INT64_MAX +
               (timeout.ticks != K_TICKS_FOREVER) * (ns / NSEC_PER_SEC),
     .tv_nsec = (timeout.ticks == K_TICKS_FOREVER) * (NSEC_PER_SEC - 1) +
                (timeout.ticks != K_TICKS_FOREVER) *
                        (ns - (ns / NSEC_PER_SEC) * NSEC_PER_SEC),
```

Instead of `(ns - (ns / NSEC_PER_SEC) * NSEC_PER_SEC)` you could simply do
`(ns % NSEC_PER_SEC)`. When computing `(ns / NSEC_PER_SEC)` earlier, the
modulus often comes for free as a byproduct of the division and the compiler
would just use it given the opportunity. Granted, a smart compiler will
conclude that the above code can be substituted by a modulus already,
meaning that in the end it doesn't make much difference to the resulting
binary, therefore it is best to just use the simpler code.

No need to respin the PR just for this hence my +1, but if you need to
re-push then this change should be considered.

cfriedt commented:
* added cast to `time_t` from `INT64_MAX` (should resolve issue [here](https://github.com/zephyrproject-rtos/zephyr/actions/runs/15144349098/job/42576485034?pr=90060))
* used modulo operator instead of multiply / divide

npitre reviewed (dismissed):
Still missing one modulo in case you repush.

cfriedt commented:
@npitre - jeez - this is like Where's Waldo 😆 Can't sneak anything past you (not that I was trying, it was just hard for me to see).

cfriedt commented:
@andyross, @peter-mitsis - please (re-)review when you have a moment. One of you must approve as assignee.

kartben commented:
It's actually also failing with gcc https://github.com/zephyrproject-rtos/zephyr/actions/runs/15197606643

cfriedt commented on include/zephyr/sys/timeutil.h:
```diff
@@ -301,6 +311,414 @@ int timeutil_sync_local_from_ref(const struct timeutil_sync_state *tsp,
+	return (k_timeout_t){
+		.ticks = ((ts->tv_sec == INT64_MAX) && (ts->tv_nsec == NSEC_PER_SEC - 1)) *
+				 K_TICKS_FOREVER +
+			 ((ts->tv_sec != INT64_MAX) && (ts->tv_sec >= 0)) *
```
> I'm seeing compile errors on this code.

This doesn't match what is in `main` currently. There was a fox submitted. Can you please ensure you are testing a commit that includes d3ec9167.
